### DISCUSSION TITLE
Simple header file cannot be parsed with CppParser component

### DISCUSSION THREAD

**Discussion by cngzhnp** [pocoproject/poco#4418] [Q&A]
**Describe the bug**
I have a simple header file (named "example.h") which needs to be parsed in order to get function name and type like below:

```cpp
#pragma once

void func1(int32_t a);
```

**To Reproduce**
To reproduce this problem only I need to write down such an example:

```cpp
void printFunctionInfo(const Poco::CppParser::Function *pFunc)
{
    std::cout << "Function name: " << pFunc->name() << '\n';
    std::cout << "Return type: " << pFunc->getReturnParameter() << '\n';
    std::cout << "Parameters:\n";
    for (auto parameter = pFunc->begin(); parameter != pFunc->end(); ++parameter)
    {
        std::cout << "Type: " << (*parameter)->declType() << ", Name: " << (*parameter)->name() << '\n';
    }
    std::cout << '\n';
}

int main(int argc, char *argv[])
{
    std::ifstream                           headerFile("example.h");
    Poco::CountingInputStream               counting_str(headerFile);
    Poco::CppParser::NameSpace::SymbolTable symbols;
    Poco::CppParser::Parser                 parser(symbols, "example.h", counting_str);

    parser.parse();

    std::cout << "Symbol size: " << symbols.size() << '\n';

    for (const auto &symbol : symbols)
    {
        if (symbol.second->kind() == Poco::CppParser::Symbol::SYM_FUNCTION)
        {
            printFunctionInfo(static_cast<Poco::CppParser::Function *>(symbol.second));
        }
    }
}
```

**Expected behavior**
I would expect that symbol size must be greater than zero but it always zero. At least I would expect to see that function name and type after call `printFunctionInfo` function.

**Logs**

**Screenshots**

**Please add relevant environment information:**

-- [cmake] Build for OS type:      Linux
-- [cmake] Build for OS version:   6.1.21.2-microsoft-standard-WSL2+
-- [cmake] Build for CPU type:     x86_64
-- [cmake] Build type:             release
-- [cmake] Build with cxx flags:   -O3 -DNDEBUG 
-- [cmake] Build with c flags:     -O3 -DNDEBUG 
-- Poco package version: 1.13.0

**Additional context**

**Comment by andrewauclair** [comment]
I'm not familiar with this part of Poco, but have you checked that `std::ifstream headerFile("example.h");` actually opens the file?

**Comment by cngzhnp** [comment]
@andrewauclair Thanks for your comment. Yes, file exists and file path is correct, during the debug session I can observe. 

The part that I could not understand, why I need to give two different objects for `Parser` class. The second parameter needs a file name as `std::string`. Therefore, third one contains `std::istream` object. In this case, I do not understand why I need for second parameter which could be extracted from third parameter.

**Comment by andrewauclair** [comment]
The parameter is a `std::istream`, not `std::ifstream`. That said, even `std::ifstream` and family don't provide access to that information.

**Comment by cngzhnp** [comment]
Yes you are right, but as I said does not make sense to me to just send `filename` and its `std::ifstream` object together as a parameter in a same object. There should be something different file that needs to be provided but I could not find it yet.

**Comment by aleks-f** [comment]
this is not released code, hence not a bug. I'l move it to discussions, and if you want to send a fix, you're welcome to do it

**Comment by cngzhnp** [comment]
After digging into a little bit, I understood that how it works. So, the second parameter was for a file name which needs to be given to `Parser` class.

However, third parameter `std::istream` should be given as `Poco::CountingInputStream` class which needs to see during exception which line causes a problem. Also, Poco::CountingInputStream must take std::ifstream which have to contain **precompiled version of header file**.

In this case, you need to generate precompiled version of the file with compiler "-E" option then you need to give it, now it works. However, naming convention, unit tests and description of this component has to be reviewed again, because without digging into, people could not understand  the workflow.
